### PR TITLE
Add crash reporting debugger event 

### PR DESCRIPTION
**Pull request by alexanderkyte**
This should help get information on native crashes that happen in hard-to-debug environments. 

To quote https://github.com/mono/mono/issues/10157 : 

When a mono runtime is attached to a managed debugger and the debugged runtime is induced to cause a crash, the state of that crash is not communicated to the managed debugger. This, combined with the difficulty in having the debugged runtime both attached to the managed debugger and lldb, makes this type of crash difficult to interactively debug.

When the managed debugger is the only connection someone has with the runtime because it is embedded in a way that is resistant to unmanaged debugging (watchOS application on-device, for instance), this is an even more difficult problem.

By adding an SDB protocol event that sends a type of "runtime death" and a blob of opaque json (version dependent, no protocol guarantees), people can report these crashes without having to assist a mono runtime engineer in observing a crash on device.

In the case that spurious bugs in this manner are submitted, it will require less effort as well to spot stack frames that indicate misuse of mixed managed-unmanaged APIs (which offer the ability to drive a segfault or g_assert

### CODE REVIEW

**Comment by alexanderkyte** [comment]
Commits moved from https://github.com/mono/mono/pull/10532

**Comment by alexanderkyte** [comment]
Hold off on giving this a review yet. I want to see what CI comes back with for this. I've still got some commented out stuff and the remnants of a few debugging helpers. I think I should split these commits into a few more with better comments too.

File: mono/mini/debugger-agent.h [hunk 1]
```diff
@@ -44,4 +44,7 @@ mono_debugger_agent_stub_init (void);
 MONO_API gboolean
 mono_debugger_agent_transport_handshake (void);
 
+void
+mono_debugger_agent_send_crash (char *jsonDump, MonoStackHash *hashes, int pauseTime);
+
```
Comment (vargaz) [inline mono/mini/debugger-agent.h]: This needs to use the debugger callback interface.

File: mono/mini/debugger-agent.c [hunk 2]
```diff
@@ -4847,6 +4856,35 @@ ss_clear_for_assembly (SingleStepReq *req, MonoAssembly *assembly)
...
+	EventInfo ei;
+
+	if (!agent_config.enabled)
+		return;
+
+	// It doesn't make sense to wait for lldb/gdb to finish if we're not
+	// actually enabled. Therefore we do the wait here.
+	sleep (pause);
+
+	mono_loader_lock ();
```
Comment (vargaz) [inline mono/mini/debugger-agent.c]: This needs to check that protocol version.

Comment (alexanderkyte) [reply mono/mini/debugger-agent.c]: I was about to ask where the downgrade happened, thanks

Comment (alexanderkyte) [reply mono/mini/debugger-agent.c]: The check was added

**Comment by alexanderkyte** [comment]
Seeing some failures here when rebasing on master. Seems caused by other code:

```
* Assertion at ../../mono/utils/refcount.h:44, condition `refcount' not met
```

**Comment by alexanderkyte** [comment]
@akoeplinger I think that the CI script is retrying the "Crash" step before it finishes on the slower lanes

**Comment by alexanderkyte** [comment]
@monojenkins commit apidiff

**Comment by alexanderkyte** [comment]
@monojenkins commit csproj

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 3]
```diff
@@ -1354,6 +1377,17 @@ class PacketWriter {
 									exit_code = r.ReadInt ();
 								//EventHandler.VMDeath (req_id, 0, null);
 								events [i] = new EventInfo (etype, req_id) { ExitCode = exit_code };
+							} else if (kind == EventKind.CRASH) {
+								ulong hash = (ulong) r.ReadLong ();
+								string dump = r.ReadString ();
+
+								// Throw here, otherwise we can be killed by VMDisconnectedExceptions 
+								// when trying to handle this event
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: This looks like some leftover comments.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachine.cs [hunk 4]
```diff
@@ -808,6 +808,12 @@ class EventHandler : MarshalByRefObject, IEventHandler
 				case EventType.UserLog:
 					l.Add (new UserLogEvent (vm, req_id, thread_id, ei.Level, ei.Category, ei.Message));
 					break;
+				case EventType.Crash:
+					// This class subclasses VMDisconnectedException, so people catching that
+					// event will cleanly handle this one. They'll just miss out on the stacktrace.
+					// throw new VMCrashException (ei.Dump, ei.Hash);
+					l.Add (new CrashEvent (vm, req_id, thread_id, ei.Dump, ei.Hash));
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachine.cs]: Ditto.

File: mcs/class/Mono.Debugger.Soft/Test/dtest-app.cs [hunk 5]
```diff
@@ -319,6 +319,10 @@ public static void wait_one ()
 			unhandled_exception_endinvoke ();
 			return 0;
 		}
+		if (args.Length >0 && args [0] == "crash_vm") {
+			crash ();
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Test/dtest-app.cs]: Should be "crash-vm" for consistency.

File: mcs/class/Mono.Debugger.Soft/Test/dtest-app.cs [hunk 6]
```diff
@@ -1330,6 +1334,23 @@ class Class3 {
 		unhandled_exception_endinvoke_2 ();
 	}
 
+	// kill(2)
+	//    int kill(pid_t pid, int sig);
+	//[DllImport (LIBC, SetLastError=true, EntryPoint="kill")]
+	//private static extern int sys_kill (int pid, int sig);
+
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Test/dtest-app.cs]: Leftover code.

File: mcs/class/Mono.Debugger.Soft/Test/dtest-app.cs [hunk 7]
```diff
@@ -1330,6 +1334,23 @@ class Class3 {
...
+	[MethodImplAttribute (MethodImplOptions.NoInlining)]
+	public static void crash () {
+		//var monoType = Type.GetType ("Mono.Runtime", false);
+		//var m = monoType.GetMethod("AbortWithTelemetry", BindingFlags.NonPublic | BindingFlags.Static);
+		//m.Invoke(null, new object[] { });
+		//var pid = Mono.Unix.UnixProcess.GetCurrentProcess ();
+		//pid.Kill ();
+		//GC.GetGeneration (10);
+		//Mono.Unix.Native.Syscall.kill (pid.Id, Mono.Unix.e.Signum.SIGABRT);
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Test/dtest-app.cs]: Leftover code.

File: mcs/class/Mono.Debugger.Soft/Test/dtest.cs [hunk 8]
```diff
@@ -2454,6 +2454,42 @@ Event step_out_await (string method, Event e)
...
+				var crash = e as CrashEvent;
+				if (crash == null)
+					continue;
+
+				success = true;
+				Assert.AreNotEqual (0, crash.Dump.Length);
+				Assert.AreNotEqual (0, crash.Hash);
+
+				Console.WriteLine ("This is the Dump: {0}", crash.Dump);
+
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Test/dtest.cs]: No logging needed.

File: mono/mini/debugger-agent.c [hunk 9]
```diff
@@ -9685,6 +9747,7 @@ mono_debugger_agent_init (void)
 	cbs.user_break = debugger_agent_user_break;
 	cbs.debug_log = debugger_agent_debug_log;
 	cbs.debug_log_is_enabled = debugger_agent_debug_log_is_enabled;
+	cbs.send_crash = mono_debugger_agent_send_crash;
```
Comment (vargaz) [inline mono/mini/debugger-agent.c]: Needs a stub in debugger-agent-stubs.c.

File: mono/mini/debugger-agent.h [hunk 10]
```diff
@@ -28,6 +28,7 @@ struct _MonoDebuggerCallbacks {
 	void (*user_break) (void);
 	void (*debug_log) (int level, MonoStringHandle category, MonoStringHandle message);
 	gboolean (*debug_log_is_enabled) (void);
+	void (*send_crash) (char *jsonDump, MonoStackHash *hashes, int pauseTime);
 };
```
Comment (vargaz) [inline mono/mini/debugger-agent.h]: The parameter names should follow the runtime coding conventions.

File: mono/mini/mini-posix.c [hunk 11]
```diff
@@ -1084,12 +1084,21 @@ dump_native_stacktrace (const char *signal, void *ctx)
...
+			mono_runtime_printf_err ("\nDebug info from gdb:\n");
 			mono_gdb_render_native_backtraces (crashed_pid);
 			exit (1);
 		}
 
-		mono_runtime_printf_err ("\nDebug info from gdb:\n");
 		waitpid (pid, &status, 0);
+
+#if !defined(DISABLE_CRASH_REPORTING)
+		// We've already done our gdb dump and our telemetry steps. Before exiting,
```
Comment (vargaz) [inline mono/mini/mini-posix.c]: The send_crash () function already contains an #if !defined(DISABLE_CRASH_REPORTING)

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachine.cs (hunk 4 above)
Comment (alexanderkyte) [reply mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachine.cs]: Removed

File: mcs/class/Mono.Debugger.Soft/Test/dtest-app.cs (hunk 5 above)
Comment (alexanderkyte) [reply mcs/class/Mono.Debugger.Soft/Test/dtest-app.cs]: 👍

File: mono/mini/debugger-agent.c (hunk 9 above)
Comment (alexanderkyte) [reply mono/mini/debugger-agent.c]: Added the stub

File: mono/mini/debugger-agent.h (hunk 10 above)
Comment (alexanderkyte) [reply mono/mini/debugger-agent.h]: It's now snakecase

File: mono/mini/mini-posix.c (hunk 11 above)
Comment (alexanderkyte) [reply mono/mini/mini-posix.c]: Okay, relying on that one now

**Comment by alexanderkyte** [comment]
It seems to be reporting that the reason for failure is ```
driver.c:7:10: fatal error: 'mono/metadata/assembly.h' file not found
#include <mono/metadata/assembly.h>```

**Comment by vargaz** [comment]
It says:

Processing triple: wasm32-wasm32-unknown-unknown
wasm.mk:104: recipe for target '/mnt/jenkins/workspace/test-mono-pull-request-wasm/sdks/builds/wasm-cross-release/wasm32-unknown-none.h' failed
emscripten include directory /mnt/jenkins/workspace/test-mono-pull-request-wasm/sdks/builds/toolchains/emsdk/emscripten/1.38.11/system/include/libc does not exist.
make: *** [/mnt/jenkins/workspace/test-mono-pull-request-wasm/sdks/builds/wasm-cross-release/wasm32-unknown-none.h] Error 1
make: *** Waiting for unfinished jobs....

**Comment by alexanderkyte** [comment]
Good catch @vargaz

File: mono/mini/debugger-agent-stubs.c [hunk 12]
```diff
@@ -91,6 +91,12 @@ stub_debugger_agent_breakpoint_from_context (MonoContext *ctx)
 	g_assert_not_reached ();
 }
 
+static void
+mono_debugger_agent_send_crash (char *json_dump, MonoStackHash *hashes, int pause)
+{
+	g_assert_not_reached ();
+}
```
Comment (vargaz) [inline mono/mini/debugger-agent-stubs.c]: Shouldn't this be empty ?

Comment (alexanderkyte) [reply mono/mini/debugger-agent-stubs.c]: I can make it empty. Some of these stubs are empty and some have the ```g_assert_not_reached ();``` , it wasn't really clear what the style was.

Comment (alexanderkyte) [reply mono/mini/debugger-agent-stubs.c]: Made it empty @vargaz

**Comment by akoeplinger** [comment]
@monojenkins commit apidiff

**Pull request merged**